pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java

```diff
@@ -266,14 +292,52 @@ public void addSegment(String segmentName, TableConfig tableConfig, IndexLoading
         manager =
             new LLRealtimeSegmentDataManager(realtimeSegmentZKMetadata, tableConfig, this, _indexDir.getAbsolutePath(),
                 indexLoadingConfig, schema, llcSegmentName, _partitionIdToSemaphoreMap.get(streamPartitionId),
-                _serverMetrics);
+                _serverMetrics, _tableUpsertMetadataManager);
       }
       _logger.info("Initialize RealtimeSegmentDataManager - " + segmentName);
       _segmentDataManagerMap.put(segmentName, manager);
       _serverMetrics.addValueToTableGauge(_tableNameWithType, ServerGauge.SEGMENT_COUNT, 1L);
     }
   }
 
+  private boolean isUpsertEnabled() {
+    return _upsertMode == UpsertConfig.Mode.FULL || _upsertMode == UpsertConfig.Mode.PARTIAL;
+  }
+
+  @Override
+  public void addSegment(ImmutableSegment immutableSegment) {
+    if (isUpsertEnabled()) {
+      handleUpsert(immutableSegment);
+    }
+    super.addSegment(immutableSegment);
+  }
+
+  private void handleUpsert(ImmutableSegment immutableSegment) {
```

Comment:
Do you suggest creating a `RealTimeUpsertTableDataManager` in `TableDataManagerProvider.getTableDataManager`?  It's an option to me, though I feel uspert shall be better built as a first-class citizen in realtime table. @Jackie-Jiang what do you think?

Reply:
Yes, I would prefer creating a `RealtimeUpsertTableDataManager` if upsert is on for a table. This can be a subclass of RealtimeTableDataManager. It will still be a first class citizen, just  keeping code independent with minimum `if` statement

Reply:
@mcvsubbu It is a little bit hard to create this new class without duplicating code. Going through the code, there are very few checks for upsert, so I think for now it is okay to colocate the upsert within the `RealtimeTableManager`. Upsert is just a feature on top of the LLC table, and it does not rely on any external services to work (different from the previous design).